server/src/main/java/org/opensearch/index/ShardIndexingPressure.java

```diff
@@ -0,0 +1,334 @@
+
+        long requestStartTime = System.nanoTime();
+        ShardIndexingPressureTracker tracker = getShardIndexingPressureTracker(shardId);
+        long nodeReplicaBytes = currentReplicaBytes.addAndGet(bytes);
+        long shardReplicaBytes = tracker.getReplicaOperationTracker().getStatsTracker().incrementCurrentBytes(bytes);
+
+        boolean shardLevelLimitBreached = false;
+        if (forceExecution == false) {
+            boolean nodeLevelLimitBreached = memoryManager.isReplicaNodeLimitBreached(tracker, nodeReplicaBytes);
+            if (nodeLevelLimitBreached == false) {
+                shardLevelLimitBreached = memoryManager.isReplicaShardLimitBreached(tracker, nodeReplicaBytes, requestStartTime);
+            }
+
+            if (shouldRejectRequest(nodeLevelLimitBreached, shardLevelLimitBreached)) {
+                replicaRejections.getAndIncrement();
+                currentReplicaBytes.addAndGet(-bytes);
+                tracker.getReplicaOperationTracker().getStatsTracker().incrementCurrentBytes(-bytes);
+                rejectShardRequest(tracker, bytes, nodeReplicaBytes, shardReplicaBytes,
+                    tracker.getReplicaOperationTracker().getRejectionTracker(), "replica");
+            }
+        }
+        totalReplicaBytes.addAndGet(bytes);
+
+        StatsTracker statsTracker = tracker.getReplicaOperationTracker().getStatsTracker();
+        markShardOperationStarted(statsTracker, tracker.getReplicaOperationTracker().getPerformanceTracker());
+        boolean isShadowModeBreach = shardLevelLimitBreached;
+
+        return () -> {
+            currentReplicaBytes.addAndGet(-bytes);
+            markShardOperationComplete(bytes, requestStartTime, isShadowModeBreach, tracker.getReplicaOperationTracker());
+            memoryManager.decreaseShardReplicaLimits(tracker);
+            tryReleaseTracker(tracker);
+        };
+    }
+
+    private boolean shouldRejectRequest(boolean nodeLevelLimitBreached, boolean shardLevelLimitBreached) {
+        return nodeLevelLimitBreached ||
+            (shardLevelLimitBreached && shardIndexingPressureSettings.isShardIndexingPressureEnforced());
+    }
+
+    private void markShardOperationStarted(StatsTracker statsTracker, PerformanceTracker performanceTracker) {
+        statsTracker.incrementRequestCount();
+        performanceTracker.incrementTotalOutstandingRequests();
+    }
+
+    private void adjustPerformanceUponCompletion(long bytes, long requestStartTime, StatsTracker statsTracker,
+                                                 PerformanceTracker performanceTracker) {
+        long requestEndTime = System.nanoTime();
+        long requestLatency = TimeUnit.NANOSECONDS.toMillis(requestEndTime - requestStartTime);
+
+        performanceTracker.addLatencyInMillis(requestLatency);
+        performanceTracker.updateLastSuccessfulRequestTimestamp(requestEndTime);
+        performanceTracker.updateTotalOutstandingRequests(0);
+
+        if(requestLatency > 0) {
```

Comment:
Although ``nanoTime`` is expected to be monotonically increasing, there are few known reasons where it can be observed backwards. Since in ``calculateRequestThroughput``, we use this value for ratio calculation, a negative value is completely undesirable, and hence this safe check.